### PR TITLE
Fixing issue on linking user module libraries with recent GCC versions

### PR DESCRIPTION
**Pull request by jgarridoalcazar**
This changes are intended to fix issue #869.

It explicitely defines "-Wl,--no-as-needed" linking flag on nest executable, nest library and pynest library. By default, recent versions of GCC set --as-needed flag. This option allows the linker to ignore, i.e., not link against, some of the libraries supplied on its command line if they are not actually used by the shared library that is being created. This option avoids external modules to be linked in NEST even if they are defined with -Dexternal-modules=....

By setting --no-as-needed GCC will link against all the libraries explicitely indicated (including external modules). It may produce slower executable/library startup if more libraries are added than required.

This PR requires PR #870 to be added in order to fix issue #868.

This PR has only been tested with GCC 5.4.0 and Ubuntu 16.04. It should be tested on different compilers and OS.

### CODE REVIEW

File: nest/CMakeLists.txt [hunk 1]
```diff
@@ -31,10 +31,16 @@ if ( IS_BLUEGENE AND static-libraries )
         )
 endif ()
 
+set_target_properties( nest
+    PROPERTIES
+    LINK_FLAGS "-Wl,--no-as-needed"
```
Comment (hakonsbm) [inline nest/CMakeLists.txt]: Remove trailing space.

File: nest/CMakeLists.txt [hunk 2]
```diff
@@ -31,10 +31,16 @@ if ( IS_BLUEGENE AND static-libraries )
...
+set_target_properties( nest
+    PROPERTIES
+    LINK_FLAGS "-Wl,--no-as-needed" 
+    )
+
 add_library( nest_lib ${nest_sources} )
 set_target_properties( nest_lib
     PROPERTIES
     OUTPUT_NAME nest
+    LINK_FLAGS "-Wl,--no-as-needed"
```
Comment (hakonsbm) [inline nest/CMakeLists.txt]: Remove trailing space here too.

**Comment by jgarridoalcazar** [comment]
Thanks @hakonsbm for reviewing this PR. The trailling spaces have already been removed.

File: nest/CMakeLists.txt [hunk 3]
```diff
@@ -31,10 +31,16 @@ if ( IS_BLUEGENE AND static-libraries )
         )
 endif ()
 
+set_target_properties( nest
+    PROPERTIES
+    LINK_FLAGS "-Wl,--no-as-needed"
```
Comment (heplesser) [inline nest/CMakeLists.txt]: This flag is not supported by the Apple linker, so you need to protect this by an `if ( NOT APPLE ) ... endif()`.  I am also not sure about the linkers for more special systems, such as BlueGene and K. Maybe the best thing would be to add a test whether `--no-as-needed` is supported by the linker and only apply it if it is. Unfortunately, this does not seem straightforward in CMake, see https://public.kitware.com/pipermail/cmake/2016-January/062623.html; we need to support older CMake versions, back to 2.8.12.

File: nest/CMakeLists.txt [hunk 4]
```diff
@@ -31,10 +31,16 @@ if ( IS_BLUEGENE AND static-libraries )
...
+    )
+
 add_library( nest_lib ${nest_sources} )
 set_target_properties( nest_lib
     PROPERTIES
     OUTPUT_NAME nest
+    LINK_FLAGS "-Wl,--no-as-needed"
```
Comment (heplesser) [inline nest/CMakeLists.txt]: See above.

File: pynest/CMakeLists.txt [hunk 5]
```diff
@@ -46,8 +46,9 @@ if ( HAVE_PYTHON )
     endif ()
     python_add_module( pynestkernel ${pynestkernel_generated_file} pynestkernel.pxd )
     if ( APPLE )
-      set_target_properties( pynestkernel PROPERTIES LINK_FLAGS "-undefined dynamic_lookup" )
+      set_target_properties( pynestkernel PROPERTIES LINK_FLAGS "-undefined dynamic_lookup -Wl,--no-as-needed" )
```
Comment (heplesser) [inline pynest/CMakeLists.txt]: See above, leave the Apple-case unchanged.

**Comment by jgarridoalcazar** [comment]
You are right, @heplesser . Removing the --no-as-needed flag in OSX systems should not be an issue since you already tested this bug in OSX and it was not reproducible (see the mailing list). After the last commit the flag is only added to non-Apple systems. It could be very helpful if anyone having access to BlueGene and/or K could test the building in these systems.

**Comment by jougs** [comment]
I have just tested the branch on BlueGene/Q (JUQUEEN) and it worked flawlesly. @heplesser: can you please confirm that it now also works on OS X so we can merge? Thanks!

**Comment by heplesser** [comment]
@jgarridoalcazar Works like a charm on macOS as well now (requires #870 merged, which is in place now). Thanks!

**Pull request merged**